src/main/java/io/cdap/plugin/cdc/source/sqlserver/ResultSetToDMLRecord.java

```diff
@@ -40,57 +40,120 @@
+        tableInformation.getName()))
       .set(Schemas.PRIMARY_KEYS_FIELD, Lists.newArrayList(tableInformation.getPrimaryKeys()))
-      .set(Schemas.OP_TYPE_FIELD, operationType.name())
+      .set(Schemas.OP_TYPE_FIELD, getChangeOperation(row).name())
       .set(Schemas.UPDATE_SCHEMA_FIELD, changeSchema.toString())
-      .set(Schemas.UPDATE_VALUES_FIELD, getChangeData(row, changeSchema))
+      .set(Schemas.UPDATE_VALUES_FIELD, map)
+      .set(Schemas.CHANGE_TRACKING_VERSION, row.getString("CHANGE_TRACKING_VERSION"))
       .build();
   }
 
-  private static Map<String, Object> getChangeData(ResultSet resultSet, Schema changeSchema) throws SQLException {
+  private static OperationType getChangeOperation(ResultSet row) throws Exception {
+    String operation = row.getString("SYS_CHANGE_OPERATION");
+    switch (operation) {
+      case "I":
+        return OperationType.INSERT;
+      case "U":
+        return OperationType.UPDATE;
+      case "D":
+        return OperationType.DELETE;
+    }
+    throw new IllegalArgumentException(String.format("Unknown change operation '%s'", operation));
+  }
+
+  private static Map<String, Object> getChangeData(ResultSet resultSet, Schema changeSchema,
+                                                   int size) throws Exception {
     ResultSetMetaData metadata = resultSet.getMetaData();
     Map<String, Object> changes = new HashMap<>();
     for (int i = 0; i < changeSchema.getFields().size(); i++) {
-      int column = i + CHANGE_TABLE_COLUMNS_SIZE;
+
+      Schema.Field field = changeSchema.getFields().get(i);
+      int column = getColumnForFeild(metadata, field.getName());
       int sqlType = metadata.getColumnType(column);
+      String sqlTypeName = metadata.getColumnTypeName(column);
       int sqlPrecision = metadata.getPrecision(column);
       int sqlScale = metadata.getScale(column);
-      Schema.Field field = changeSchema.getFields().get(i);
-      Object sqlValue = DBUtils.transformValue(sqlType, sqlPrecision, sqlScale, resultSet, field.getName());
-      Object javaValue = transformSQLToJavaType(sqlValue);
-      changes.put(field.getName(), javaValue);
+      getColumnForFeild(metadata, field.getName());
+
+      try {
+        /**
+         * Handling clob and blob data type ... JTDS does not support free() that throws exception from the DBUtils.
+         */
+        if (sqlType == 2005) {
+          Clob clob = resultSet.getClob(field.getName());
+          String retVal = (clob != null ? clob.getSubString(1, (int) clob.length()) : null);
+          changes.put(field.getName(), "\"" + retVal.toString()  + "\"");
+        } else if (sqlType == 2004) {
+          Blob blob = resultSet.getBlob(field.getName());
+          byte[] blobVal = (blob != null ? blob.getBytes(1L, (int) blob.length()) : null);
+          changes.put(field.getName(), "\"" + blobVal.toString()  + "\"");
+        } else if (sqlType == -8) {
```

Comment:
what are these types? Please add comments about what this is doing.

Reply:
These are blob and clob data types, jtds has issues with reading these types so we had to add this

Reply:
this looks really weird... why is toString() being called on every single value and getting wrapped in quotes? This is for sure breaking behavior and also means data types are lost.